protocols/relay/src/v2/client.rs

```diff
@@ -0,0 +1,412 @@
+        _: &ConnectedPoint,
+        _failed_addresses: Option<&Vec<Multiaddr>>,
+    ) {
+        self.connected_peers
+            .entry(*peer_id)
+            .or_default()
+            .push(*connection_id);
+    }
+
+    fn inject_connection_closed(
+        &mut self,
+        peer_id: &PeerId,
+        connection_id: &ConnectionId,
+        _: &ConnectedPoint,
+        _handler: handler::Handler,
+    ) {
+        self.connected_peers.get_mut(peer_id).map(|cs| {
+            cs.remove(
+                cs.iter()
+                    .position(|c| c == connection_id)
+                    .expect("Connection to be known."),
+            )
+        });
+    }
+
+    fn inject_event(
+        &mut self,
+        event_source: PeerId,
+        _connection: ConnectionId,
+        handler_event: handler::Event,
+    ) {
+        match handler_event {
+            handler::Event::ReservationReqAccepted { renewal } => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::ReservationReqAccepted {
+                            relay_peer_id: event_source,
+                            renewal,
+                        },
+                    ))
+            }
+            handler::Event::ReservationReqFailed { renewal } => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::ReservationReqFailed {
+                            relay_peer_id: event_source,
+                            renewal,
+                        },
+                    ))
+            }
+            handler::Event::OutboundCircuitReqFailed {} => {
+                self.queued_actions
+                    .push_back(NetworkBehaviourAction::GenerateEvent(
+                        Event::OutboundCircuitReqFailed {
+                            relay_peer_id: event_source,
+                        },
+                    ))
+            }
+            handler::Event::InboundCircuitReqDenied { src_peer_id } => self
+                .queued_actions
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    Event::InboundCircuitReqDenied { src_peer_id },
+                )),
+            handler::Event::InboundCircuitReqDenyFailed { src_peer_id, error } => self
+                .queued_actions
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    Event::InboundCircuitReqDenyFailed { src_peer_id, error },
+                )),
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        _poll_parameters: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        if let Some(event) = self.queued_actions.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        loop {
+            match self.from_transport.poll_next_unpin(cx) {
```

Comment:
I don't really understand what the `loop` is for? We return from every branch, plus on `Pending` we `break` and return `Pending`. That is exactly what `ready!` does.

```suggestion
	let msg = match ready!(self.from_transport.poll_next_unpin(cx)) {
		Some(msg) => msg,
		None => unreachable!(...) 
	};
	
	let action = match msg {
		// map to NetworkBehaviourAction::{Dial,NotifyHandler}
	};

	return Poll::Ready(action);
```

Reply:
Well spotted. Thanks. Fixed in 7a4cf0a970ea7c9fd5b4d89dd0c809accb7057c0.